### DISCUSSION TITLE
configuration issue with the Prometheus metrics exporter

### DISCUSSION THREAD

**Discussion by mitht** [open-telemetry/opentelemetry-java-instrumentation#11144] [Q&A] [answered]
opentelemetry-javaagent  version: 1.33.1
prometheus               version: 2.51.1.darwin-amd64
grafana                  version: 10.4.1

# environment variable
OTEL_EXPORTER_OTLP_ENDPOINT=http://127.0.0.1:4317;OTEL_INSTRUMENTATION_COMMON_DB_STATEMENT_SANITIZER_ENABLED=false;OTEL_METRICS_EXPORTER=prometheus;OTEL_EXPORTER_PROMETHEUS_HOST=127.0.0.1;OTEL_EXPORTER_PROMETHEUS_PORT=9090

# javaagent
-javaagent:/app/opentelemetry-javaagent-1.33.1.jar -Dotel.service.name=my-service -Dotel.instrumentation.jdbc.enabled=true -Dotel.logs.exporter=none

After configuring OTEL_METRICS_EXPORTER=prometheus;OTEL_EXPORTER_PROMETHEUS_HOST=127.0.0.1;OTEL_EXPORTER_PROMETHEUS_PORT=9090, Grafana fails to connect to Prometheus, with the following error message:

<img width="1108" alt="image" src="https://github.com/open-telemetry/opentelemetry-java-instrumentation/assets/29425860/6c34d401-b6a5-4bb4-ac51-9dfe46917663">

Without Prometheus environment variables configured, visit: http://127.0.0.1:9090/graph

<img width="1123" alt="image" src="https://github.com/open-telemetry/opentelemetry-java-instrumentation/assets/29425860/6cc0dd23-d32c-4b68-b106-d8a59f2301d6">

With the environment configuration added, OTEL_METRICS_EXPORTER=prometheus;OTEL_EXPORTER_PROMETHEUS_HOST=127.0.0.1;OTEL_EXPORTER_PROMETHEUS_PORT=9090, visit: http://127.0.0.1:9090/graph.

<img width="1115" alt="image" src="https://github.com/open-telemetry/opentelemetry-java-instrumentation/assets/29425860/eba3e973-8ae5-417e-9a60-2639cc155177">

**Comment by mitht** [comment]
Error message:

ReadObject: expect { or , or } or n, but found #, error found in #1 byte of ...|# HELP http|..., bigger context ...|# HELP http_server_duration_milliseconds The durati|... - There was an error returned querying the Prometheus API.

**Comment by trask** [comment]
hi @mitht, is there any exception stack trace logged in the application that's running the java agent?

if not, can you try to repro using the OTel Java SDK + Prometheus exporter directly to help narrow down the issue?

**Reply by mitht** [reply]
Prometheus and Java program logs are normal

<img width="1824" alt="image" src="https://github.com/open-telemetry/opentelemetry-java-instrumentation/assets/29425860/ee901f9e-bdc3-4c3f-915c-24cb0db5c4ec">


After adding the environment configuration OTEL_METRICS_EXPORTER=prometheus;OTEL_EXPORTER_PROMETHEUS_HOST=127.0.0.1;OTEL_EXPORTER_PROMETHEUS_PORT=9090 and starting the Java program, Grafana cannot connect to Prometheus.

<img width="1122" alt="image" src="https://github.com/open-telemetry/opentelemetry-java-instrumentation/assets/29425860/9e4dc086-0569-4ccf-ab58-b75090b6a85e">


grafana logs:

{"@level":"warn","@message":"Failed to get prometheus buildinfo","@timestamp":"2024-04-17T10:56:32.175530+08:00","err":"failed to unmarshal JSON: invalid character '#' looking for beginning of value","logger":"tsdb.prometheus"}
{"@level":"warn","@message":"Failed to get prometheus heuristics","@timestamp":"2024-04-17T10:56:32.175623+08:00","err":"failed to get buildinfo: failed to unmarshal JSON: invalid character '#' looking for beginning of value","logger":"tsdb.prometheus"}
INFO [04-17|10:56:32] Request Completed                        logger=context userId=1 orgId=1 uname=admin method=GET path=/api/datasources/uid/edisdw2bm1b0gd/health status=400 remote_addr=127.0.0.1 time_ms=75 duration=75.867773ms size=261 referer=http://127.0.0.1:3000/connections/datasources/edit/edisdw2bm1b0gd handler=/api/datasources/uid/:uid/health status_source=server
INFO [04-17|10:56:40] Completed cleanup jobs                   logger=cleanup duration=46.949668ms


<img width="1829" alt="image" src="https://github.com/open-telemetry/opentelemetry-java-instrumentation/assets/29425860/a9780176-8886-40ff-be0a-82b37ab07d46">


After removing these configurations, access is normal: OTEL_METRICS_EXPORTER=prometheus; OTEL_EXPORTER_PROMETHEUS_HOST=127.0.0.1; OTEL_EXPORTER_PROMETHEUS_PORT=9090

<img width="1108" alt="image" src="https://github.com/open-telemetry/opentelemetry-java-instrumentation/assets/29425860/60878462-8346-463f-8b8e-8a4d21c83093">

**Comment by laurit** [comment]
cc @zeitlinger

**Comment by mitht** [comment, marked as answer]
The issue is resolved by adding the following configuration to Prometheus.
The port 9090 should not be used. 

- job_name: "otelcol"
    static_configs:
      - targets: ["localhost:17890"]
